RequestHandler.php

```diff
@@ -0,0 +1,253 @@
+<?php
+
+namespace PHPPM;
+
+use React\Socket\UnixConnector;
+use React\Socket\TimeoutConnector;
+use React\Socket\ConnectionInterface;
+
+class RequestHandler
+{
+    /**
+     * @var ConnectionInterface
+     */
+    private $incoming;
+
+    /**
+     * @var ConnectionInterface
+     */
+    private $connection;
+
+    /**
+     * @var ProcessManager
+     */
+    private $processManager;
+
+    /**
+     * @var Slave instance
+     */
+    private $slave;
+
+    private $connectionOpen = true;
+    private $redirectionTries = 0;
+    private $incomingBuffer = '';
+
+    public function __construct(ProcessManager $processManager)
+    {
+        $this->processManager = $processManager;
+
+        // substitute properties below
+        $this->output = $processManager->output;
+        $this->loop = $processManager->loop;
+        $this->maxRequests = $processManager->maxRequests;
+        $this->timeout = $processManager->timeout;
+    }
+
+    public function handle(ConnectionInterface $incoming)
+    {
+        $this->incoming = $incoming;
+
+        $this->incoming->on('data', [$this, 'handleData']);
+        $this->incoming->on('close', function () {
+            $this->connectionOpen = false;
+        });
+
+        $this->start = microtime(true);
+        $this->getNextSlave();
+    }
+
+    /**
+     * Buffer incoming data until slave connection is available
+     * and headers have been received
+     */
+    public function handleData($data)
+    {
+        $this->incomingBuffer .= $data;
+
+        if ($this->connection && $this->isHeaderEnd($this->incomingBuffer)) {
+            $remoteAddress = $this->incoming->getRemoteAddress();
+            $headersToReplace = [
+                'X-PHP-PM-Remote-IP' => trim(parse_url($remoteAddress, PHP_URL_HOST), '[]'),
+                'X-PHP-PM-Remote-Port' => trim(parse_url($remoteAddress, PHP_URL_PORT), '[]')
+            ];
+
+            $buffer = $this->replaceHeader($this->incomingBuffer, $headersToReplace);
+            $this->connection->write($buffer);
+
+            $this->incoming->removeListener('data', [$this, 'handleData']);
+            $this->incoming->pipe($this->connection);
+        }
+    }
+
+    /**
+     * Get next free slave from process manager.
+     * Asynchronously keep trying until slave becomes available
+     */
+    public function getNextSlave()
+    {
+        $slave =& $this->processManager->getNextSlave();
```

Comment:
I'd remove that reference and array stuff, and replace it by a POPO instead

Reply:
> I'd remove that reference and array stuff, and replace it by a POPO instead. Then we could have nice is*() methods as well instead of manual checking array values everywhere.

Absolutely. Plus add something like a `SlavePool` that manages the connection. I've noticed though that- in order for me to do this- I need the code much smaller and more digestible.

So this is only a step in that direction and will need+see some further work.